# **Binary, Multi-Class and Multi-Label Classification**

# **Practice**

# **Multi-Label Tweet Post Classification**

Adriana, a student of a data science crash course, is learning different types of classifications as a pat of text preprocessing through the NLTK library of Python. She is asked to explore a collection of nearly 27,480 posted tweets for model training purposes and 3,533 posted tweets from the Kaggel site. She is also asked to build clasifiers to test with a set of unknown documents as part of her learning Help Adriana learn text classification using multi-class classification machine learning methods.

Data Structure

Data is split between thetrain and test sets and based on tweet posts.

Data Set Characteristics:

text: the text of the tweet

sentiment: the sentiment or polarity of the tweet


In [ ]:
import numpy as np                                       # for numerical calculation
import pandas as pd                                      # for data analysis
import re                                                # regular expression
import json                                              # java script file
import warnings
warnings.filterwarnings('ignore')                        # to ignore warnings
import string                                            # string
from nltk.corpus import stopwords                        # stop words

from sklearn.feature_extraction.text import TfidfVectorizer #Tfidvectorizer
from sklearn.linear_model import LogisticRegression      # classification model
from sklearn.multiclass import OneVsRestClassifier
from sklearn.metrics import confusion_matrix,classification_report

# **Task 1**


Load the tweet corpus train and test datasets. Perform the missing value treatme it if required.


In [ ]:
df_train = pd.read_csv("DS3_C3_S2_TweetTrain_Data_Challenge.csv") # train data
df_test = pd.read_csv("DS3_C3_S2_TweetTest_Data_Challenge.csv")   # test data
df_train.head()

textID                                               text  \
0  cb774db0d1                I`d have responded, if I were going   
1  549e992a42      Sooo SAD I will miss you here in San Diego!!!   
2  088c60f138                          my boss is bullying me...   
3  9642c003ef                     what interview! leave me alone   
4  358bd9e861   Sons of ****, why couldn`t they put them on t...   

                         selected_text sentiment Time of Tweet Age of User  \
0  I`d have responded, if I were going   neutral       morning        0-20   
1                             Sooo SAD  negative          noon       21-30   
2                          bullying me  negative         night       31-45   
3                       leave me alone  negative       morning       46-60   
4                        Sons of ****,  negative          noon       60-70   

       Country  Population -2020  Land Area (Km�)  Density (P/Km�)  
0  Afghanistan          38928346         652860.0               60  
1      Albania           2877797          27400.0              105  
2      Algeria          43851044        2381740.0               18  
3      Andorra             77265            470.0              164  
4       Angola          32866272        1246700.0               26

In [ ]:
df_train.isnull().sum() #train data has some null value need to treat the missing value

textID              0
text                1
selected_text       1
sentiment           0
Time of Tweet       0
Age of User         0
Country             0
Population -2020    0
Land Area (Km�)     0
Density (P/Km�)     0
dtype: int64

In [ ]:
df_train.dropna(inplace=True) # droping the missing value

In [ ]:
df_train.isnull().sum() # now there is no null value

textID              0
text                0
selected_text       0
sentiment           0
Time of Tweet       0
Age of User         0
Country             0
Population -2020    0
Land Area (Km�)     0
Density (P/Km�)     0
dtype: int64

In [ ]:
df_test.head() # test data

textID                                               text sentiment  \
0  f87dea47db  Last session of the day  http://twitpic.com/67ezh   neutral   
1  96d74cb729   Shanghai is also really exciting (precisely -...  positive   
2  eee518ae67  Recession hit Veronique Branquinho, she has to...  negative   
3  01082688c6                                        happy bday!  positive   
4  33987a8ee5             http://twitpic.com/4w75p - I like it!!  positive   

  Time of Tweet Age of User      Country  Population -2020  Land Area (Km�)  \
0       morning        0-20  Afghanistan          38928346         652860.0   
1          noon       21-30      Albania           2877797          27400.0   
2         night       31-45      Algeria          43851044        2381740.0   
3       morning       46-60      Andorra             77265            470.0   
4          noon       60-70       Angola          32866272        1246700.0   

   Density (P/Km�)  
0               60  
1              105  
2               18  
3              164  
4               26

In [ ]:
df_test.isnull().sum() # no null value in test data

textID              0
text                0
sentiment           0
Time of Tweet       0
Age of User         0
Country             0
Population -2020    0
Land Area (Km�)     0
Density (P/Km�)     0
dtype: int64

# **Task 2**

Preprocess the train and test tweet texts using the following two methods:

punctuation

Method #1: Cleaning the reviews - removal of stopwords and import string

def text_process(text):

nopunc = [char for char in text if char not in string.punctuation]

nopunc = ".join(nopunc)

return [w ord for word in nopunc.split() if word.lower() not in stopwords.words('english')]

Method #2: Conversion of review text from list type to string type def listToString(s): str1="" return (str1.join(s))


In [ ]:
# cleaning the reviews- removal of stopwords and punctuation
def text_process(text):
    nopunc = [char for char in text if char not in string.punctuation]
    nopunc = ''.join(nopunc)
    return[word for word in nopunc.split() if word.lower() not in stopwords.words("english")]

In [ ]:
# conversion of review text form list type to string type data
def listToString(s):
    str1 = " "
    return (str1.join(s))

In [ ]:
train=df_train['text'].apply(text_process) #cleaning

In [ ]:
train=train.apply(listToString) # converting list to string

In [ ]:
train

0                                       Id responded going
1                                  Sooo SAD miss San Diego
2                                            boss bullying
3                                    interview leave alone
4                 Sons couldnt put releases already bought
                               ...                        
27476    wish could come see u Denver husband lost job ...
27477    Ive wondered rake client made clear NET dont f...
27478    Yay good Enjoy break probably need hectic week...
27479                                                worth
27480                   flirting going ATG smiles Yay hugs
Name: text, Length: 27480, dtype: object

In [ ]:
test=df_test['text'].apply(text_process)

In [ ]:
test=test.apply(listToString)

In [ ]:
test

0                    Last session day httptwitpiccom67ezh
1       Shanghai also really exciting precisely skyscr...
2       Recession hit Veronique Branquinho quit compan...
3                                              happy bday
4                                httptwitpiccom4w75p like
                              ...                        
3529                            3 im tired cant sleep try
3530    alone old house Thanks net keeps alive kicking...
3531    know mean little dog sinking depression wants ...
3532           sutra next youtube video gonna love videos
3533           httptwitpiccom4woj2 omgssh ang cute ng bby
Name: text, Length: 3534, dtype: object

# Task 3

1. Convert text data into vector form.

2. Classify data of four classes using a one-vs-rest strategy with Logistic Regression.


In [ ]:
myvector = TfidfVectorizer()
myobject = myvector.fit_transform(train)
myobject1 = myvector.transform(test)

In [ ]:
myobject.toarray() # vector format

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

2)One Vs Rest Classifier on Multi_lable

In [ ]:
LR = LogisticRegression(max_iter=100) # model
ovr = OneVsRestClassifier(LR)
model = ovr.fit(myobject,df_train['sentiment'])
y_pred = model.predict(myobject1)

In [ ]:
model.score(myobject,df_train['sentiment']) # accuracy of train data

0.7846069868995633

In [ ]:
model.score(myobject1,df_test['sentiment']) # accuracy of test data

0.6983588002263724

# Task 4

1. Print the first 30 samples of predicted and actual output as a table

2. Display the Confusion Matrix & classificatio. report

3. Write your interpretation based on your observations.


In [ ]:
data = {"Predicted":df_test['sentiment'][:30],"Actual":y_pred[:30]}
df_output = pd.DataFrame(data)
df_output # printing output in data frame

Predicted    Actual
0    neutral   neutral
1   positive  positive
2   negative  negative
3   positive  positive
4   positive  positive
5   positive  positive
6   negative  negative
7   negative  negative
8    neutral   neutral
9    neutral   neutral
10  negative  negative
11   neutral   neutral
12   neutral   neutral
13  negative  negative
14  negative  negative
15  negative  negative
16  negative  negative
17  negative  negative
18  positive  positive
19   neutral  positive
20  negative  negative
21  negative  negative
22  negative  negative
23   neutral   neutral
24  negative   neutral
25  positive  positive
26   neutral   neutral
27   neutral   neutral
28  positive   neutral
29  negative  negative

In [ ]:
confusion_matrix(df_test['sentiment'],y_pred) # confusion matrix

array([[ 621,  340,   40],
       [ 195, 1069,  166],
       [  34,  291,  778]], dtype=int64)

In [ ]:
print(classification_report(df_test['sentiment'],y_pred)) # classification report

              precision    recall  f1-score   support

    negative       0.73      0.62      0.67      1001
     neutral       0.63      0.75      0.68      1430
    positive       0.79      0.71      0.75      1103

    accuracy                           0.70      3534
   macro avg       0.72      0.69      0.70      3534
weighted avg       0.71      0.70      0.70      3534



Most of the prediction are almost same as test data of accuarcy 78%